src/Serializer/PayloadSerializer.php

```diff
@@ -215,12 +215,32 @@ private function serializeAsEnvelope(Event $event): string
             }
         }
 
-        $itemHeader = [
+        $eventItemHeader = [
             'type' => (string) $event->getType(),
             'content_type' => 'application/json',
         ];
 
-        return sprintf("%s\n%s\n%s", JSON::encode($envelopeHeader), JSON::encode($itemHeader), $this->serializeAsEvent($event));
+        $items = [
+            JSON::encode($envelopeHeader),
+            JSON::encode($eventItemHeader),
+            $this->serializeAsEvent($event),
+        ];
+
+        foreach ($event->getAttachments() as $attachment) {
+            $data = $attachment->getData();
+
+            $attachmentItemHeader = [
+                'type' => 'attachment',
+                'filename' => $attachment->getFilename(),
+                'content_type' => $attachment->getContentType(),
+                'length' => \strlen($data),
+            ];
+
+            $items[] = JSON::encode($attachmentItemHeader);
+            $items[] = $data;
```

Comment:
As discussed earlier it might be better to only read the file here once. Otherwise we have it at least twice in memory.